### PR TITLE
Add type annotations to torch.onnx.* modules

### LINKED ISSUE #45215: Enable torch.onnx.* typechecks during CI
`TestTypeHints.test_run_mypy` should pass with `[mypy-torch.onnx.*]` deleted from mypy.ini

cc @ezyang @malfet @rgommers @xuzhao9 @gramster

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Fixes #45215 

Still need to resolve a few mypy issues before a review. In special, there is an error which I don't know how to solve, see:
```python
torch/onnx/utils.py:437: error: Name 'is_originally_training' is not defined  [name-defined]
        if training is None or training == TrainingMode.EVAL or (training == TrainingMode.PRESERVE and not is_originally_training):
```

`is_originally_training` is used but never defined/imported on [`torch/onnx/utils.py`](https://github.com/pytorch/pytorch/blob/ab5cc97fb078d936603a7039856192366ac32a5e/torch/onnx/utils.py#L437),

### CODE REVIEW

File: torch/onnx/utils.py [hunk 1]
```diff
@@ -76,7 +77,7 @@ def export(model, args, f, export_params=True, verbose=False, training=None,
     if aten or export_raw_ir:
         assert operator_export_type is None
         assert aten ^ export_raw_ir
-        operator_export_type = OperatorExportTypes.ATEN if aten else OperatorExportTypes.RAW
+        operator_export_type = OperatorExportTypes.ONNX_ATEN if aten else OperatorExportTypes.RAW
```
Comment (guilhermeleobas) [inline torch/onnx/utils.py:80]: There is no mention to `OperatorExportTypes.ATEN` in the source code. So, I think this value was supposed to be `OperatorExportTypes.ONNX_ATEN`.

File: torch/onnx/utils.py [hunk 2]
```diff
@@ -466,7 +468,7 @@ def export_to_pretty_string(model, args, f, export_params=True, verbose=False, t
...
```
Comment (guilhermeleobas) [inline torch/onnx/utils.py:471]: Same here

File: torch/onnx/utils.py [hunk 3]
```diff
@@ -432,7 +434,7 @@ def _model_to_graph(model, args, verbose=False,
     param_names = input_and_param_names[len(input_and_param_names) - len(params):]
     params_dict = dict(zip(param_names, params))
 
-    if training is None or training == TrainingMode.EVAL or (training == TrainingMode.PRESERVE and not is_originally_training):
```
Comment (guilhermeleobas) [inline torch/onnx/utils.py:435]: `is_originally_training` is not defined anywhere.

**Review by rgommers — APPROVED** [review]
LGTM, thanks @guilhermeleobas

**Comment by rgommers** [comment]
This has a merge conflict, can you rebase @guilhermeleobas?

**Comment by guilhermeleobas** [comment]
Failures are real
```
torch/onnx/utils.py:203: error: Module has no attribute "_jit_pass_onnx_set_dynamic_input_shape"  [attr-defined]
torch/onnx/utils.py:225: error: Module has no attribute "_jit_pass_onnx_graph_shape_type_inference"; maybe "_jit_pass_onnx_node_shape_type_inference"?  [attr-defined]
torch/onnx/utils.py:429: error: Module has no attribute "_jit_pass_onnx_assign_output_shape"  [attr-defined]
torch/onnx/utils.py:435: error: Module has no attribute "_jit_pass_onnx_assign_output_shape"  [attr-defined]
torch/onnx/symbolic_opset11.py:806: error: Module has no attribute "_jit_pass_fixup_onnx_loop_node_inputs"; maybe "_jit_pass_fixup_onnx_controlflow_node"?  [attr-defined]
```

**Review by malfet — APPROVED** [review]

**Comment by rgommers** [comment]
@guilhermeleobas there's still CI failures here that are real:
```
Traceback (most recent call last):
  File "test_type_hints.py", line 217, in test_run_mypy
    self.fail(f"mypy failed: {stdout} {stderr}")
AssertionError: mypy failed: torch/_C/__init__.pyi:375: error: Name 'Value' already defined on line 368  [no-redef]
torch/_C/__init__.pyi:379: error: Name 'Block' already defined on line 371  [no-redef]
torch/_C/__init__.pyi:383: error: Name 'Node' already defined on line 365  [no-redef]
Found 3 errors in 1 file (checked 1099 source files)
```
Can you update the PR?

**Comment by guilhermeleobas** [comment]
@rgommers those issues should now be fixed.

The current mypy failures are not related to any changes introduced in this PR and there are fixed (ignored) in a different [PR](https://github.com/pytorch/pytorch/pull/46490)
```
~/git/pytorch:torch-onnx (pytorch-cuda-dev) guilhermel@pytorch-dev $ mypy
torch/distributions/kl.py:107: error: Value of type variable "_LT" of "min" cannot be "_Match"  [type-var]
torch/distributions/kl.py:108: error: Value of type variable "_LT" of "min" cannot be "_Match"  [type-var]
torch/utils/cpp_extension.py:1560: error: Argument 2 to "load_module" has incompatible type "IO[Any]"; expected "Optional[_FileLike]"  [arg-type]
torch/testing/_internal/common_utils.py:1040: error: Value of type variable "_LT" of "max" cannot be "Optional[float]"  [type-var]
```

**Comment by rgommers** [comment]
> The current mypy failures are not related to any changes introduced in this PR and there are fixed (ignored) in a different [PR](https://github.com/pytorch/pytorch/pull/46490)

Yes. Would be nice to get those fixes landed first and then re-run the CI on this PR, because the `onnx_ort_test1/2` jobs would be good to see green.

**Comment by guilhermeleobas** [comment]
Failure is not related to this PR:
```
Oct 22 20:00:18 + cleanup
Oct 22 20:00:18 + retcode=0
Oct 22 20:00:18 + set +x
Oct 22 20:00:18 Error response from daemon: No such exec instance: 9d9da493a69d31cab48a6c46379347093d2a2fcf77e3a9a486be69535b1dac3d

Exited with code exit status 1
```

**Comment by rgommers** [comment]
LGTM now, thanks @guilhermeleobas.

@malfet would you be able to land this PR?

**Comment by mruberry** [comment]
Unlanding. This appears to have broken multiple builds (see the Github CI, too). Sample snippet:

```
test_run_mypy - TestTypeHints
test_type_hints.py
Traceback (most recent call last):
  File "test_type_hints.py", line 217, in test_run_mypy
    self.fail(f"mypy failed: {stdout} {stderr}")
AssertionError: mypy failed: torch/_C/__init__.pyi:281: error: Name '_jit_pass_constant_propagation' already defined on line 206  [no-redef]
torch/onnx/symbolic_helper.py:368: error: Name 'unbind' already defined (possibly by an import)  [no-redef]
Found 2 errors in 2 files (checked 1157 source files)
```

**Comment by rgommers** [comment]
Grr, I was relying on my old approval. @guilhermeleobas can you rebase and fix the duplicate definitions?

**Comment by guilhermeleobas** [comment]
@rgommers done!

**Comment by ezyang** [comment]
@guilhermeleobas could you please reopen a new PR? I cna't import this one anymore

**Comment by guilhermeleobas** [comment]
@ezyang, sure!

**Pull request closed without merging**